source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,538 @@
+    {
+        AZLogError( ( "Failed to initialize options: Invalid argument \r\n" ) );
+        ret = eAzureIoTHubClientInvalidArgument;
+    }
+    else
+    {
+        pxHubClientOptions->pucModelID = NULL;
+        pxHubClientOptions->ulModelIDLength = 0;
+        pxHubClientOptions->pucModuleID = NULL;
+        pxHubClientOptions->ulModuleIDLength = 0;
+        pxHubClientOptions->pucUserAgent = ( const uint8_t * ) azureiothubUSER_AGENT;
+        pxHubClientOptions->ulUserAgentLength = sizeof( azureiothubUSER_AGENT ) - 1;
+        ret = eAzureIoTHubClientSuccess;
+    }
+
+    return ret;
+}
+/*-----------------------------------------------------------*/
+
+AzureIoTHubClientResult_t AzureIoTHubClient_Init( AzureIoTHubClientHandle_t xAzureIoTHubClientHandle,
+                                                  const uint8_t * pucHostname,
+                                                  uint16_t ulHostnameLength,
+                                                  const uint8_t * pucDeviceId,
+                                                  uint16_t ulDeviceIdLength,
+                                                  AzureIoTHubClientOptions_t * pxHubClientOptions,
+                                                  uint8_t * pucBuffer,
+                                                  uint32_t ulBufferLength,
+                                                  AzureIoTGetCurrentTimeFunc_t xGetTimeFunction,
+                                                  const AzureIoTTransportInterface_t * pxTransportInterface )
+{
+    AzureIoTHubClientResult_t ret;
+    az_iot_hub_client_options options;
+
+    if( ( xAzureIoTHubClientHandle == NULL ) ||
+        ( pucHostname == NULL ) || ( ulHostnameLength == 0 ) ||
+        ( pucDeviceId == NULL ) || ( ulDeviceIdLength == 0 ) ||
+        ( pucBuffer == NULL ) || ( ulBufferLength == 0 ) ||
+        ( xGetTimeFunction == NULL ) || ( pxTransportInterface == NULL ) )
+    {
+        AZLogError( ( "Failed to initialize: Invalid argument \r\n" ) );
+        ret = eAzureIoTHubClientInvalidArgument;
+    }
+    else if ( ( ulBufferLength < azureiotTOPIC_MAX ) ||
+              ( ulBufferLength < ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) ) )
+    {
+        AZLogError( ( "Failed to initialize: Not enough memory passed \r\n" ) );
+        ret = eAzureIoTHubClientOutOfMemory;
+    }
+    else
+    {
+        memset( ( void * ) xAzureIoTHubClientHandle, 0, sizeof( AzureIoTHubClient_t ) );
+
+        /* Setup scratch buffer to be used by middleware */
+        xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength =
+            ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) > azureiotTOPIC_MAX ? ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) : azureiotTOPIC_MAX;
+        xAzureIoTHubClientHandle->_internal.pucAzureIoTHubClientWorkingBuffer = pucBuffer;
+        pucBuffer += xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength;
```

Comment:
Could we create a new variable for `pucBuffer` so intent later on is more clear?  I'm making up (bad) name but like below.

This keeps pucBuffer which is passed the same, and any reasonable compiler optimizer won't have trouble re-using register for `pucBuffer` and a new variable `pucNetworkBuffer`.

```c
uint8_t * pucNetworkBuffer = pucBuffer + xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength;
// Then pass the pucNetworkBuffer to AzureIoTMQTT_Init
```

Reply:
agreed. changing